Add zephyr,display chosen property and use it to obtain default display controller in the samples

jfischer-no opened the pull request:
This allows us to make samples that use the display (controller) API more generic and remove ifdefery like
`#define DISPLAY_DEV_NAME DT_LABEL(DT_INST(0, ilitek_ili9340))`.

This patch adds bindings for the display controller integrated in a display, like SSD1306.
We also have devices in the tree where panel and controller are separate entities and
are connected via a defined interface, see mimxrt10xx boards. These are described
differently in devicetree. It may make sense to unify it even if it seems a bit to excessive
for device like SSD1306.
However it is about how the drivers are configured, application that uses display (controller)
API is not affected much since the API provides an interface to retrieve the runtime information,
except it is LVGL adaption layer. It is not possible completely remove options like
LVGL_DISPLAY_DEV_NAME without heavy reworking of LVGL adaption layer
(or complete removal of LVGL from zephyr tree, for which I immediately give +1),
because lvgl_init needs the name of the device, and LVGL_*_RES_MAX options to define
internal buffers.

jfischer-no committed:
drivers: ssd16xx: fix driver initialization

ssd16xx driver is not well designed and does not pass configuration
via device->config but via struct ssd16xx_data, this was not taken
into account in the commit 4d6d50e2bc60
("display: ssd16xx: convert to spi_dt_spec").

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards/shields: add zephyr,display chosen property

Add zephyr,display chosen property.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dts: add bindings for display controller

Add bindings for display controller, where a controller
is a fixed part of the display.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
samples: use chosen property to get default display controller

Rework samples to use DEVICE_DT_GET and DT_CHOSEN
to get default display controller device.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: display_sdl: rework to obtain configuration from devicetree

Add bindings for SDL based display controller.
Rework SDL based display controller driver to obtain
configuration from devicetree. Remove unnecessary casts,
add multi-instance support.

Add display controller node and chosen property
to native_posix devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: display_dummy: rework to obtain configuration from devicetree

Add bindings for dummy display controller.
Rework dummy display controller driver to obtain
configuration from devicetree. Remove unnecessary casts,
add multi-instance support.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
boards: remove Kconfig option LVGL_DISPLAY_DEV_NAME

Remove Kconfig option LVGL_DISPLAY_DEV_NAME.
Also remove it from mimxrt10xx_evk boards to have consistent
configuration, even if they do not benefit from it that much.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
lib: lvgl: obtain LVGL_DISPLAY_DEV_NAME from devicetree

It is not possible completely remove options like
LVGL_DISPLAY_DEV_NAME without heavy reworking of LVGL support
because lvgl_init needs the name of the device, and
LVGL_*_RES_MAX options to define internal buffers.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
dts: display: include common display controller bindings

Include common display controller bindings.
Add required height and width properties to "ilitek,ili9340"
and "ilitek,ili9488" users.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
drivers: display: remove unused Kconfig options

remove unused options SDL_DISPLAY_* and DUMMY_DISPLAY_*.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

gmarull reviewed (approved):
Thanks for these changes, nice cleanup

jfischer-no committed:
tests: build_all: remove display controller nodes from sensor test

Remove unused display controller nodes from sensor test.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

mbolivar-nordic reviewed (changes requested):
This needs to be documented in the list of chosen nodes

mbolivar-nordic commented on dts/bindings/display/ilitek,ili9xxx-common.yaml:
```diff
@@ -4,7 +4,7 @@
 description: ILI9XXX display controllers common properties.
 
-include: spi-device.yaml
+include: [spi-device.yaml, display-controller.yaml]
```
This seems to break bisectability by adding required `height` and `width` properties to the following bindings without updating in-tree users:

- dts/bindings/display/ilitek,ili9340.yaml
- dts/bindings/display/ilitek,ili9488.yaml

The dts/bindings/display/ilitek,ili9341.yaml binding is also affected, but doesn't seem to have any in-tree users.

jfischer-no replied:
yes, realized it later, I will do better.

mbolivar-nordic commented on boards/arm/bl5340_dvk/bl5340_dvk_cpuapp_common.dts:
```diff
@@ -195,14 +196,16 @@
 		reset-gpios = <&gpio0 6 GPIO_ACTIVE_LOW>;
 		cmd-data-gpios = <&gpio0 12 GPIO_ACTIVE_LOW>;
 		rotation = <270>;
+		width = <240>;
+		height = <320>;
```
I think these sorts of changes should be put into the previous patch which added the properties to the binding, for bisectability.

mbolivar-nordic commented on samples/drivers/display/src/main.c:
```diff
@@ -211,13 +180,12 @@ void main(void)
+	if (!device_is_ready(display_dev)) {
 		LOG_ERR("Device %s not found. Aborting sample.",
-			DISPLAY_DEV_NAME);
+			DT_LABEL(DT_CHOSEN(zephyr_display)));
```
`display_dev->name`. Here and elsewhere in this patch, getting the name from the device struct seems a little cleaner.

mbolivar-nordic commented on boards/shields/ls0xx_generic/Kconfig.defconfig:
```diff
@@ -13,10 +13,6 @@ config LS0XX
-	string "Display device name"
-	default "LS0XX"
-
 config LVGL_HOR_RES_MAX
 	int "Horizantal resolution"
 	default 128
```
Do you also need a separate patch removing these? These values should be coming from devicetree now, right?

jfischer-no replied:
Yes, except for mimxrt10xx boards.

mbolivar-nordic reviewed (changes requested):
Thanks very much for this cleanup. I've got a few more comments now that I had time to do a more detailed review; my earlier -1 was just to keep this from being merged too quickly.

jfischer-no commented on drivers/display/ssd16xx.c:
```diff
@@ -636,8 +636,8 @@ static int ssd16xx_controller_init(const struct device *dev)
 static int ssd16xx_init(const struct device *dev)
 {
-	const struct ssd16xx_config *config = dev->config;
 	struct ssd16xx_data *driver = dev->data;
+	const struct ssd16xx_config *config = driver->config;
```
not just moving around, `= dev->config;` vs. `= driver->config;`
it is part of https://github.com/zephyrproject-rtos/zephyr/pull/39708

mbolivar-nordic replied:
Ah, thanks! I missed that.

keith-zephyr reviewed (approved):
LGTM pending Marti's requested changes.

jfischer-no committed:
boards: remove options LVGL_HOR_RES_MAX and LVGL_VER_RES_MAX

Remove options LVGL_HOR_RES_MAX and LVGL_VER_RES_MAX.
Keep the options for mimxrt10xx_evk boards as they
cannot yet be obtained there from devicetree.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no commented on drivers/display/display_dummy.c:
```diff
@@ -4,9 +4,14 @@
+#include <device.h>
+
+#define DUMMY_DISPLAY_X_RES	DT_INST_PROP(0, width)
+#define DUMMY_DISPLAY_Y_RES	DT_INST_PROP(0, height)
```
I was lazy, reworked both, display_dummy.c and display_sdl.c, to support multi-instance.

mbolivar-nordic commented on samples/subsys/display/cfb_custom_font/src/main.c:
```diff
@@ -11,23 +11,20 @@
 		printk("Display device not ready\n");
 	}
 
+	if (display_set_pixel_format(display, PIXEL_FORMAT_MONO10) != 0) {
```
This appears to be an unrelated change for this commit; what's going on here?

jfischer-no replied:
sloppiness, will update after https://github.com/zephyrproject-rtos/zephyr/pull/40876 is merged.

mbolivar-nordic replied:
#40876 doesn't seem to have anything to do with this; is that the right link?

jfischer-no replied:
yes, if I push now, CI will fail again because of that.

jfischer-no committed:
samples: cfb_custom_font: set required pixel format

Set for this sample required pixel format.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

jfischer-no committed:
doc: devicetree: add "zephyr,display" to chosen nodes table

Add "zephyr,display" to Zephyr-specific chosen nodes table.

Signed-off-by: Johann Fischer <johann.fischer@nordicsemi.no>

mbolivar-nordic commented:
` samples: crypto: exclude CTR mode for mbedtls `

Did you mean to include this commit in this PR?

jfischer-no commented:
> `samples: crypto: exclude CTR mode for mbedtls`
> 
> Did you mean to include this commit in this PR?

rather to test what happens,
https://github.com/zephyrproject-rtos/zephyr/pull/40765/checks?check_run_id=4447344836
most of the time I can fix the bugs that CI uncovers myself, but here actions log offers no help at all.

mbolivar-nordic commented:
> rather to test what happens,

Not really sure if that's a yes or a no :). But  #40876 is merged now, so I'm guessing you plan on making one more update to this PR before we can merge it, right?

jfischer-no commented:
> > rather to test what happens,
> 
> Not really sure if that's a yes or a no :). But #40876 is merged now, so I'm guessing you plan on making one more update to this PR before we can merge it, right?

Actually, nothing more is planned.  What annoys me is only the unstable CI.

mbolivar-nordic commented:
> nothing more is planned

OK, I see now you did split the pixel format change into its own commit, but you just didn't mention it. LGTM, thanks.

jfischer-no commented:
rebased to include 474d6866d02d5b74efe9e991c9ed7db5f436ca5b
